### ISSUE TITLE
entity insert order is not expected

### ISSUE THREAD

**Issue by kzhui125** [dotnet/efcore#30310]
For the following code, I expect post1 is inserted before post2. but actually, post2 is inserted before post1, this is not expected.

```cs
blog.Posts.Add(post1);
blog.Posts.Add(post2);
db.Add(blog);
await db.SaveChangesAsync();
```

full code is as below:

Program.cs:

```cs
using System;
using System.Reflection;
using Microsoft.EntityFrameworkCore;


using (var db = new BloggingContext())
{
    Blog blog = new() { Url = "blog1" };
    
    GoodPost post1 = new()
    {
        Title = "Good1",
        Content = "Good1 Content",
        GoodData = "Good1 Data",
    };

    BadPost post2 = new()
    {
        Title = "Bad1",
        Content = "Bad1 Content",
        BadData = "Bad1 Data",
    };

    blog.Posts.Add(post1);
    blog.Posts.Add(post2);
    db.Add(blog);
    await db.SaveChangesAsync();
    Console.WriteLine();
}


public class BloggingContext : DbContext
{
    public DbSet<Blog> Blogs => Set<Blog>();
    public DbSet<PostBase> Posts => Set<PostBase>();

    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
    {
        string connectionString = "some connection string";
        optionsBuilder.UseSqlServer(connectionString);
    }

    protected override void OnModelCreating(ModelBuilder builder)
    {
        builder.ApplyConfigurationsFromAssembly(Assembly.GetExecutingAssembly());
        base.OnModelCreating(builder);
    }
}

public class Blog
{
    public int Id { get; set; }
    public required string Url { get; set; }
    public List<PostBase> Posts { get; set; } = new();
}

public abstract class PostBase
{
    public int Id { get; set; }
    public required string Title { get; set; }
    public required string Content { get; set; }

    public int BlogId { get; set; }
    public Blog Blog { get; set; } = null!;
}

public class GoodPost : PostBase
{
    public required string GoodData { get; set; }
}

public class BadPost : PostBase
{
    public required string BadData { get; set; }
}
```

PostConfiguration.cs:

```cs
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Metadata.Builders;

public class PostConfiguration : IEntityTypeConfiguration<PostBase>
{
    public void Configure(EntityTypeBuilder<PostBase> builder)
    {
        builder.UseTphMappingStrategy()
            .ToTable("Posts");

        builder.Property(p => p.Title)
            .HasMaxLength(200);

        builder.HasDiscriminator<string>("PostType")
            .HasValue<GoodPost>("Good")
            .HasValue<BadPost>("Bad");

        builder.Property("PostType")
            .HasMaxLength(100);
    }
}

public class PostAConfiguration : IEntityTypeConfiguration<GoodPost>
{
    public void Configure(EntityTypeBuilder<GoodPost> builder)
    {
        builder.Property(p => p.GoodData)
            .HasMaxLength(1000);
    }
}

public class PostBConfiguration : IEntityTypeConfiguration<BadPost>
{
    public void Configure(EntityTypeBuilder<BadPost> builder)
    {
        builder.Property(p => p.BadData)
            .HasMaxLength(1000);
    }
}
```

this is the executed sql, you can see the INSERT statement of post1 is before post2:

```sql
exec sp_executesql N'SET NOCOUNT ON;
INSERT INTO [Posts] ([BadData], [BlogId], [Content], [PostType], [Title])
OUTPUT INSERTED.[Id]
VALUES (@p1, @p2, @p3, @p4, @p5);
INSERT INTO [Posts] ([BlogId], [Content], [GoodData], [PostType], [Title])
OUTPUT INSERTED.[Id]
VALUES (@p6, @p7, @p8, @p9, @p10);
',N'@p1 nvarchar(1000),@p2 int,@p3 nvarchar(4000),@p4 nvarchar(100),@p5 nvarchar(200),@p6 int,@p7 nvarchar(4000),@p8 nvarchar(1000),@p9 nvarchar(100),@p10 nvarchar(200)',@p1=N'Bad1 Data',@p2=1,@p3=N'Bad1 Content',@p4=N'Bad',@p5=N'Bad1',@p6=1,@p7=N'Good1 Content',@p8=N'Good1 Data',@p9=N'Good',@p10=N'Good1'
```

**Comment by ajcvickers** [comment]
@kzhui125 This is by-design. EF does not preserve the order that entities are added to the context. Further, EF then creates an ordering for inserts, updates, and deletes based on a variety of factors. Beyond this, the order that entities are returned by a database can be different from the order they are inserted, and can change from query to query unless an explicit OrderBy is used.

**Comment by kzhui125** [comment]
does this mean I can't rely on the order even with OwnsMany (which is auto included when querying the owner)?

```cs
public class Distributor
{
    public int Id { get; set; }
    public List<SomeClass> Items { get; set; }
}

modelBuilder.Entity<Distributor>().OwnsMany(
    p => p.Items
```

with OwnsMany, Do I have to write `.Include(x => x.Items.OrderBy(x => x.OrderId))` to guarantee the order?

This should be well documented(for example in OwnsMany).

Thanks.

**Comment by roji** [comment]
@kzhui125 in relational databases, the insertion order general does not matter - you must always specify an ordering when querying if you wish the results to be sorted. So even if EF generated the insert order you want, that still wouldn't guarantee that your included items come back ordered by their ID. You must explicitly have `OrderBy(x => x.OrderId`) as you've written above.

**Comment by kzhui125** [comment]
Thanks @roji 

for the following code example:

```cs
using System;
using System.Reflection;
using Microsoft.EntityFrameworkCore;

using (var context = new MyContext())
{
    Distributor d1 = new()
    {
        Items = new()
        {
            new ChildItem { Name = "Item1" },
            new ChildItem { Name = "Item2" },
            new ChildItem { Name = "Item3" },
        },
    };
    context.Add(d1);
    await context.SaveChangesAsync();
    Console.WriteLine();
}

using (var context = new MyContext())
{
    var distributors = context.Distributors
        .ToList();
    Console.WriteLine();
}


public class MyContext : DbContext
{
    public DbSet<Distributor> Distributors => Set<Distributor>();

    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
    {
        string connectionString = File.ReadAllText("./ConnectionString.txt");
        optionsBuilder.UseSqlServer(connectionString);
    }

    protected override void OnModelCreating(ModelBuilder builder)
    {
        builder.ApplyConfigurationsFromAssembly(Assembly.GetExecutingAssembly());
        base.OnModelCreating(builder);
    }
}

public class Distributor
{
    public int Id { get; set; }
    public List<ChildItem> Items { get; set; }
}

public class ChildItem
{
    public int Id { get; set; }
    public required string Name { get; set; }
}
```

```cs
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Metadata.Builders;

public class DistributorConfiguration : IEntityTypeConfiguration<Distributor>
{
    public void Configure(EntityTypeBuilder<Distributor> builder)
    {
        builder.ToTable("Distributors");

        builder.OwnsMany(p => p.Items, i =>
        {
            i.ToTable("Distributor_Items");
        });
    }
}
```

the generated insert sql is

```sql
exec sp_executesql N'SET IMPLICIT_TRANSACTIONS OFF;
SET NOCOUNT ON;
MERGE [Distributor_Items] USING (
VALUES (@p0, @p1, 0),
(@p2, @p3, 1),
(@p4, @p5, 2)) AS i ([DistributorId], [Name], _Position) ON 1=0
WHEN NOT MATCHED THEN
INSERT ([DistributorId], [Name])
VALUES (i.[DistributorId], i.[Name])
OUTPUT INSERTED.[Id], i._Position;
',N'@p0 int,@p1 nvarchar(4000),@p2 int,@p3 nvarchar(4000),@p4 int,@p5 nvarchar(4000)',
@p0=1,@p1=N'Item1',
@p2=1,@p3=N'Item2',
@p4=1,@p5=N'Item3'
```

the generated query sql for `context.Distributors.ToList();` is (for which items is auto included):

I think it would be better if  `[d0].[DistributorId]` is changed to `[d0].[Id]`, sinse it is a duplicate of `[d].[id]`
```sql
SELECT [d].[Id], [d0].[DistributorId], [d0].[Id], [d0].[Name]
FROM [Distributors] AS [d]
LEFT JOIN [Distributor_Items] AS [d0] ON [d].[Id] = [d0].[DistributorId]
ORDER BY [d].[Id], [d0].[DistributorId]
```

So There are 3 options now:

1. EFCore sql generation change `[d0].[DistributorId]` to `[d0].[Id]`, when using `context.Distributors.ToList()`,  can I rely the order then?
2. use `context.Distributors.Include(d => d.Items.OrderBy(i => i.Id)).ToList()` instead of `context.Distributors.ToList()`, does this works?
3. add a `OrderId` field to `ChildItem` class, then use `OrderBy(x => x.OrderId)`, It works but I will have to write more code.

> So even if EF generated the insert order you want, that still wouldn't guarantee that your included items come back ordered by their ID.

Can you tell that wouldn't guarantee is from `database` side or `ef core` side? Does the database respect the insert sql item order? (in this example is Item1/Item2/Item3)

If it is the database can't guarantee, I think I have to use option 3.

Thanks a lot.

**Comment by roji** [comment]
> I think it would be better if [d0].[DistributorId] is changed to [d0].[Id], sinse it is a duplicate of [d].[id]

I might be misunderstanding, but that doesn't seem to be true. `[d0].[DistributorId]` is the foreign key column in the (dependent) `Distributor_Items` table, which points to the `[Id]` column on the (principal) `Distributors` table. This is what causes the databases to match up the correct child items to each distributor.

Stepping back... When querying, you can *never* rely on any ordering, ever, unless your SQL contains an explicit ORDER BY clause which enforces it; that's just how databases work.

Now, trying to control ordering via ID is generally not feasible; as you can see, in the above SQL EF uses an advanced MERGE statement to insert mulitple rows in a single statement - that's quite important to performance. That also means that there's no guarantee as to which row gets which ID: parallelization on the server may cause Item3 to get a lower ID than Item1 (at least in theory).

So yes, if you want to control the ordering in which something comes back, yuo need to manage your own ordering column (e.g. add OrderId to ChildItem) and explicitly order by it in your queries.

What you're looking is probably #9067, which would make EF manage this for you.

**Comment by roji** [comment]
Duplicate of #9067

**Comment by AloysA** [comment]
@roji In particular circumstance the order of the inserts (and/or updates) may matter for eliminating deadlocks in highly concurrent systems. Is there any EF Core functionality available to control the order or do we need to revert to generating our own SQL in this case?

**Comment by roji** [comment]
@AloysA EF applies deterministic ordering inside SaveChanges precisely in order to prevent deadlocks - that's precisely one of the reasons why you can't rely on your Add() invocations to be performed in-order. Are you seeing a scenario where EF's SaveChanges is causing a deadlock? If so, that's something we need to be able to reproduce and fix on our side, rather than you using SQL.

**Comment by AloysA** [comment]
@roji We do in parallel some updates and insert into the same table (where the inserts use `.AddRange()`). These interact with each other and cause a deadlock. They deadlock on each other, so the inserts by themselves work fine in parallel and the updates as well. It is the combination that causes issues. We have verified by implementing our own "bulk insert" with the consistent ordering, for our use case, that the deadlocks disappear.

In this case it could be helpful if EF Core would support `AddRange<T>(IEnumerable<T> entities, bool? preserveOrdering = false)` or perhaps it should be more generic to support this for the whole context rather than on the Add/AddRange methods. Maybe something like `dbContext.DbSet.SetInsertOrderBy(entity => entity.PropertyX)`?

**Comment by roji** [comment]
> We do in parallel some updates and insert into the same table (where the inserts use .AddRange()). These interact with each other and cause a deadlock.

As I wrote above, that's exactly the scenario that EF's SaveChanges is supposed to prevent, by applying consistent, deterministic ordering on the SQL statements it sends to the database. And that's why I'm asking for a minimal repro from you; if you're able to produce a deadlock purely by doing two SaveChanges concurrently, then this may indicate a bug in EF itself which we need to investigate and fix.

In other words, the way forward isn't for you to manually tweak the ordering of INSERTs/UPDATEs sent by SaveChanges, but rather for SaveChanges to always order things correctly so a deadlock isn't produced (that's the way EF and SaveChanges work).

So I'm hoping you can help us track this down by providing a minimal, runnable code sample that shows the deadlock occuring.

**Comment by AloysA** [comment]
For context, we are storing system statistics in our solution, aggregating and updating up to 500.000 metrics per minute. For that process we have written our own optimized code and SQL statements for updates. Our consistent ordering here is by explicitly ordering by the primary key.

However, we have other parts that are less extreme where we like to stick to EF Core. For insert statements you cannot order by the primary key (would also no be an issue), but since there are some indexes we can at least order by the most important foreign key. I'm not sure how EF Core is doing consistent ordering for inserts, perhaps it is documented somewhere so that we can see if we can adopt to that approach for our updates (or not).

**Comment by roji** [comment]
@AndriySvyryd can maybe provide some info on the ordering used by SaveChanges.

**Comment by AndriySvyryd** [comment]
Sure. Though the ordering process would be hard to replicate exactly outside of EF, so for highly concurrent tables you should either only use a specific approach or be diligent to never mix inserts with updates.

1. We apply a [topological ordering](https://github.com/dotnet/efcore/blob/9c2dc753491b4a9e96981960862017264c90b770/src/EFCore.Relational/Update/Internal/CommandBatchPreparer.cs#L363) based on these constraints:
  a. Commands adding rows must precede commands adding or modifying rows that will be referencing the former
  b. Commands deleting rows or modifying the foreign key values must precede commands deleting rows or modifying the rows that are currently being referenced by the former
  c. Commands deleting rows or modifying unique constraint values must precede commands adding or modifying unique constraint values to the same values
  d. Commands deleting rows must precede commands adding rows for the same table unless this causes a dependency cycle
2. We then [sort the commands](https://github.com/dotnet/efcore/blob/9c2dc753491b4a9e96981960862017264c90b770/src/EFCore.Relational/Update/Internal/ModificationCommandComparer.cs#L20) within each group (batch) based on:
  a. The target table
  b. Type (Delete/Update/Insert)
  c. The primary key value (for inserts with a store-generated key this is usually the order in which the corresponding entities were added to the context)
